fem/particleset.hpp

```diff
@@ -0,0 +1,561 @@
+// Copyright (c) 2010-2025, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_PARTICLESET
+#define MFEM_PARTICLESET
+
+#include "../config/config.hpp"
+#include "../linalg/linalg.hpp"
+#include "gslib.hpp"
+#include "kernel_dispatch.hpp"
+
+namespace mfem
+{
+
+/// Container for data associated with a single particle
+/** See ParticleSet for more information. */
+class Particle
+{
+protected:
+   Vector coords;
+   std::vector<Vector> fields;
+   std::vector<Array<int>> tags;
```

Comment:
Why 2 dimensions? Most public functions seem to use only the first dimension.

Reply:
We used `Array<int>` here solely so that we can use `Array<T>::MakeRef`, so that a `Particle` instance can either be owning or non-owning of all its data - nothing else (each array length is strictly just 1).

We did this so that we can keep the class Rule of Zero more easily, without having to carefully define custom copy/move semantics.

Reply:
I would also keep `Array<int>`, because this will most likely need to be extended to arbitrary length in the future. I suggested above to support it on the level of `Particle` already now, which would also avoid the confusion with the extra dimension.